### ISSUE TITLE
pycbc_inspiral is unable to find bundled fftw3 libraries

### ISSUE THREAD

**Issue by paulhopkins** [gwastro/pycbc#2914]
Using the CVMFS based PyCBC virtual environment, the command `pycbc_inspiral` fails to find the bundled fftw3 libraries:

```
/cvmfs/oasis.opensciencegrid.org/ligo/sw/pycbc/x86_64_rhel_7/virtualenv/pycbc-v1.14.1/bin/pycbc_inspiral", line 28, in <module>;
    import pycbc.fft.fftw, pycbc.version
/cvmfs/oasis.opensciencegrid.org/ligo/sw/pycbc/x86_64_rhel_7/virtualenv/pycbc-v1.14.1/lib/python2.7/site-packages/pycbc/fft/fftw.py", line 35, in <module>
    raise ImportError("Unable to find FFTW libraries")
ImportError: Unable to find FFTW libraries
```

The libraries are in `$VIRTUAL_ENV/lib/python2.7/site-packages/lalsuite.libs/` but it is not sufficient to add this directory to `LD_LIBRARY_PATH` because they contain a version hash in the name. A work around is to create symlinks with basic names and add this directory to `LD_LIBRARY_PATH`:

```
mkdir .libs && cd .libs
ln -s $VIRTUAL_ENV/lib/python2.7/site-packages/lalsuite.libs/libfftw3f-*.so* libfftw3f.so
ln -s $VIRTUAL_ENV/lib/python2.7/site-packages/lalsuite.libs/libfftw3-*.so* libfftw3.so
export LD_LIBRARY_PATH=`pwd`
```

The processing scheme is set to `mkl` so if it can skip this step or not.

**Comment by spxiwh** [comment]
Hi @paulhopkins, thanks for reporting this. As I'm reading this, this sounds like an ARCCA issue and not a PyCBC issue. Am I misreading this, or are you asking for us to do something on our end here?

**Comment by paulhopkins** [comment]
Hi @spxiwh I think this is a general issue: do you see this problem on the OSG at all? I would like cvmfs Virtualenv installation of pycbc_inspiral to be able to automatically pick up the bundled libraries, much like the LALSuite libraries already do:

```
ldd /cvmfs/oasis.opensciencegrid.org/ligo/sw/pycbc/x86_64_rhel_7/virtualenv/pycbc-v1.14.1/lib/python2.7/site-packages/lal/_lal.so  | grep fftw
        libfftw3-c8dcb815.so.3.2.4 => /cvmfs/oasis.opensciencegrid.org/ligo/sw/pycbc/x86_64_rhel_7/virtualenv/pycbc-v1.14.1/lib/python2.7/site-packages/lal/../lalsuite.libs/libfftw3-c8dcb815.so.3.2.4 (0x00002ad692b04000)
        libfftw3f-044fdcc5.so.3.2.4 => /cvmfs/oasis.opensciencegrid.org/ligo/sw/pycbc/x86_64_rhel_7/virtualenv/pycbc-v1.14.1/lib/python2.7/site-packages/lal/../lalsuite.libs/libfftw3f-044fdcc5.so.3.2.4 (0x00002ad692e0c000)
```

This would also guarantee that the correct libraries are being used. I think firstly you need to add symlinks to the `lalsuite.libs` directory, e.g. `libfftw3.so -> libfftw3-c8dcb815.so.3.2.4` so that it can be found by the `get_ctypes_library` function. Secondly, you could hard-code this directory, so that you don't need to rely on `LD_LIBRARY_PATH`:

```
    libsdirs = []
    # First check for bundled libraries
    libsdirs += os.path.join(sys.path[-1], "lalsuite.libs")
    # Then try to get from LD_LIBRARY_PATH
```
I am happy to make a PR, but I don't see where the libraries get bundled into `lalsuite.libs`

**Comment by ahnitz** [comment]
@paulhopkins On OSG (I can't speak for the LSC here) we've used singularity images which package the entire environment so this isn't an issue normally. 

To be clear are you requesting that we support the lalsuite included libraries as a location for fftw3f? The standard way would be to expect that fftw3f be installed with the pk-config files so it can be located that way, but I'm not sure how fftw3 if being bundled with lalsuite on conda. 

I agree though LD_LIBRARY_PATH is not a proper answer. In general we only support it for testing purposes. 

I have no objections if the solution works to include the lalsuite.libs path on the pycbc side, but that probably means that the lalsuite package should be modified to provide ungarbled (i.e. without the random string) shared object files. I don't know if that's doable, but might be work asking the lalsuite packagers. 

@josh-willis Do you have thoughts on this issue?

**Comment by paulhopkins** [comment]
Hi @ahnitz sorry for the confusion, this issue is NOT about the Conda installation, where it is able to find it via `pkg-config`.

This is only about the CVMFS distributed Virtualenv. I see now that the `lalsuite.libs` directory is added by the `lalsuite` wheel and not by yourselves. I don't think it is necessary to ungarble the library names, but just adding a name symlink should be sufficient.

I will work up a PR. You say that you use LD_LIBRARY_PATH for testing, therefore, I will make this first on the list of locations.

**Comment by josh-willis** [comment]
@paulhopkins So, I'm still not clear what's being proposed: are you wanting to load the FFTW that lalsuite distributes as part of the CVMFS virtualenv?  I don't think that's appropriate, and if we do decide we wanted to do that, then the proper way would be to install FFTW itself into the virtualenv.  A virtual environment simply is not a conda environment or a container, and should not be expected to behave as such. If you're talking about just running on ARCCA via the virtual environment, then you should install FFTW on ARCCA, or load it as a module, for example. 

Longer term, we could consider removing FFTW as a requirement of PyCBC, and the FFT will fail if there is no underlying FFT library, but that perhaps is reasonable behavior.  Historically, we required FFTW because it is itself a requirement of lalsuite (and lalsuite is a requirement of PyCBC). But back then, the pip installable lalsuite didn't exist (or was not the version we needed), so we would build lalsuite from source into a virtual environment. Given that the pip installable lalsuite provides its own FFTW, this is arguably not needed anymore. @ahnitz what would you think about just leaving it up to users to have some FFT library, rather than failing if at least FFTW is not there?

**Comment by josh-willis** [comment]
Actually, looking again at the FFT code, I think we dropped the requirement for the underlying FFT library to have FFTW specifically already, though it will fail if you have  `scheme` of `cpu` and cannot find FFTW (but that failure should happen).  The failure Paul is seeing is because of the PyCBC inspiral executable itself.

I think the proper solution is to drop the direct import of `pycbc.fft.fftw`, though I should check that the needed argument options will still get imported by the top-level fft module whenever `fftw` is available.  Then the calls to the specific FFTW options at the beginning and end of the context block should be made conditional on finding `'fftw'` in `pycbc.fft.get_backend-names()`

**Comment by ahnitz** [comment]
@josh-willis I think also @paulhopkins is looking for pycbc to be able to pick up FFTW from the lalsuite location and not just a system installation.

**Comment by paulhopkins** [comment]
Hi @josh-willis I think what you are suggesting is best. At the moment we are specifying the MKL processing scheme, so if FFTW is not actually required then it should not fail.

Since this also affects any user installed virtualenv I have created a PR that addresses the issue of picking up the lalsuite.libs bundled libraries: #2920. If  you prefer for it to be fixed a different way then feel free to reject this.

Finally, I have been testing with Conda installed versions of PyCBC and if that can be made to work then I think that would be preferable.

**Comment by ahnitz** [comment]
Addressed by #2926.
